enhancements/installer/openstack-ipi-failure-domains.md

```diff
@@ -0,0 +1,254 @@
+
+Destroying a cluster must make sure that no resources are deleted that didn't belong to the cluster.  Leaking resources is preferred over any possibility of deleting non-cluster resources.
+
+### Workflow Description
+
+- The OpenStack administrators will deploy OpenStack with multiple Availability Zones, so Nova and Cinder can respectfully deploy servers and volumes in the Failure Domain.
+- The OpenStack administrators will create at least one Routed Provider Network, and then multiple subnets, where each zone has a least one subnet.
+- The OpenShift administrators will identify the Domain Failures: their availability zones, subnets, etc. They'll decide which ones will be used for the masters, and the ones for the workers (domains can be shared).
+- The OpenShift administrators will provide the right configuration in the `install-config.yaml` file and then deploy the cluster (detailed later in this document).
+- (Out of scope for this enhancement, see "Non-Goals") The OpenShift administrators will make sure that the network infrastructure has a networking route to reach the OCP Control Plane VIPs.
+
+### API Extensions
+
+Not applicable.
+
+
+### Implementation Details/Notes/Constraints
+
+#### Types
+
+```golang
+// OpenStackPlatformFailureDomainSpec holds the zone failure domain and
+// the Neutron subnet of that failure domain.
+type OpenStackPlatformFailureDomainSpec struct {
+  // name defines the name of the OpenStackPlatformFailureDomainSpec
+  Name        string `json:"name"`
+  ComputeZone string `json:"computeZone"`
+  StorageZone string `json:"storageZone"`
+  Subnet      string `json:"subnet"`
+}
+```
+
+#### Resource provided to the installer
+
+The OpenShift administrators provide a list of `failureDomains` and their information:
+
+```yaml
+platform:
+  openstack:
+    ...
+    failureDomains:
+    - name: rack1
+      computeZone: az1
+      storageZone: cinder-az1
+      subnet: d7ffd3d8-87e1-481f-a818-ff2f17787d40
+    - name: rack2
+      computeZone: az2
+      storageZone: cinder-az2
+      subnet: f584ce8c-bb54-422b-8e11-6a530be4c23e
+    - ...
+```
+There might be additional options in the domains, like flavors or BGP peers, but for now we will cover the minimum.
+
+Then they'll choose what Failure Domains to use when deploying the masters and workers directly in the machine pools:
+
+```yaml
+controlPlane:
+  name: master
+  platform:
+    openstack:
+      failureDomains:
+      - rack1
+      - rack2
+      - rack3
+  replicas: 3 // this will deploy 1 master in each domain
+compute:
+- name: worker
+  platform:
+    openstack:
+      ...
+      failureDomains:
+      - rack1
+      - rack2
+      - rack3
+      - rack4
+      - rack5
+      rootVolume:
+        size: 30
+        type: performance
+  replicas: 10 // this will deploy 2 workers per rack
+```
+
+The installer will create the control plane machines in a way that they can easily be adopted by the upcoming ControlPlane machineset.
```

Comment:
Can you provide a link to the ControlPlane machineset EP?  It maybe helpful for the people reviewing this EP to review or understand the CP machineset change to ensure compatibility.